calyx-opt/src/passes/compile_sync_without_sync_reg.rs

```diff
@@ -0,0 +1,211 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::{self as ir, GetAttributes, RRC};
+use calyx_ir::{build_assignments, guard, structure};
+use calyx_utils::{CalyxResult, Error};
+use std::collections::HashMap;
+
+#[derive(Default)]
+/// Compiles @sync without use of std_sync_reg
+/// Upon encountering @sync, it first instantiates a std_reg(1) for each thread(`bar`)
+/// and a std_reg(1) for each barrier (`s`)
+/// It then continuously assigns the value of (`s.in`) to 1'd1 guarded by the
+/// expression that all values of `bar` for threads under the barrier are
+/// set to 1'd1
+/// Then it replaces the @sync control operator with
+/// seq {
+///     barrier;
+///     clear;
+/// }
+/// `barrier` simply sets the value of `bar` to 1'd1 and then waits
+/// for `s.out` to be up
+/// `clear` resets the value of `bar` to 1'd0 for reuse of barrier
+/// Using this method, each thread only incurs 3 cycles of latency overhead for
+/// the barrier, and we theoretically won't have a limit for number of threads
```

Comment:
Does this mean that the pass currently only support 2 threads?

Reply:
No the pass can support as many threads as wanted.
Latency-wise the overhead is O(1), if all threads get to the barrier at the same time.

Reply:
Got it! Is there a simple test you can write to check that multiple threads work?

Reply:
Yes I'm working on that